python/ql/src/experimental/semmle/python/frameworks/Stdlib.qll

```diff
@@ -8,221 +8,64 @@ private import semmle.python.dataflow.new.DataFlow
+    DataFlow::Node regexNode;
+    Attribute regexMethod;
 
-    /**
-     * Gets a reference to the attribute `attr_name` of the `re` module.
-     * WARNING: Only holds for a few predefined attributes.
-     */
-    private DataFlow::Node re_attr(DataFlow::TypeTracker t, string attr_name) {
-      attr_name in [
-          "match", "fullmatch", "search", "split", "findall", "finditer", "sub", "subn", "compile",
-          "escape"
-        ] and
-      (
-        t.start() and
-        result = DataFlow::importNode("re" + "." + attr_name)
-        or
-        t.startInAttr(attr_name) and
-        result = re()
-      )
-      or
-      // Due to bad performance when using normal setup with `re_attr(t2, attr_name).track(t2, t)`
-      // we have inlined that code and forced a join
-      exists(DataFlow::TypeTracker t2 |
-        exists(DataFlow::StepSummary summary |
-          re_attr_first_join(t2, attr_name, result, summary) and
-          t = t2.append(summary)
-        )
-      )
+    DirectRegex() {
+      this = API::moduleImport("re").getMember(any(ReMethods m)).getACall() and
+      regexNode = this.getArg(0) and
+      regexMethod = this.asExpr().(Attribute)
     }
 
-    pragma[nomagic]
-    private predicate re_attr_first_join(
-      DataFlow::TypeTracker t2, string attr_name, DataFlow::Node res, DataFlow::StepSummary summary
-    ) {
-      DataFlow::StepSummary::step(re_attr(t2, attr_name), res, summary)
-    }
+    override DataFlow::Node getRegexNode() { result = regexNode }
 
-    /**
-     * Gets a reference to the attribute `attr_name` of the `re` module.
-     * WARNING: Only holds for a few predefined attributes.
-     */
-    private DataFlow::Node re_attr(string attr_name) {
-      result = re_attr(DataFlow::TypeTracker::end(), attr_name)
-    }
+    override Attribute getRegexMethod() { result = regexMethod }
+  }
 
-    /**
-     * Gets a reference to any `attr_name` of the `re` module that immediately executes an expression.
-     * WARNING: Only holds for a few predefined attributes.
-     */
-    private DataFlow::Node re_exec_attr() {
-      exists(string attr_name |
-        attr_name in ["match", "fullmatch", "search", "split", "findall", "finditer", "sub", "subn"] and
-        result = re_attr(DataFlow::TypeTracker::end(), attr_name)
+  private class CompiledRegex extends DataFlow::CallCfgNode, RegexExecution::Range {
+    DataFlow::Node regexNode;
+    Attribute regexMethod;
+
+    CompiledRegex() {
+      exists(DataFlow::CallCfgNode patternCall, DirectRegex reMethod |
+        this = reMethod and
+        patternCall = API::moduleImport("re").getMember("compile").getACall() and
+        patternCall = reMethod.(DataFlow::AttrRead).getObject().getALocalSource() and
+        regexNode = patternCall.getArg(0) and
+        // regexMethod is *not* worked out outside class instanciation because `CompiledRegex` focuses on re.compile(pattern).ReMethod
+        regexMethod = reMethod.getRegexMethod()
       )
     }
```

Comment:
I am not sure that it works using a `DirectRegex` here. As I understand it, you can call the methods named by `ReMethods` in two ways:
1. directly on the `re` module with a string input, this will compile the string and execute the resulting regex; this is modelled by `DirectRegex`.
2. on a compiled regex, constructed via `re.compile`; this is modelled by `CompiledRegex`. Here the regex string is found as the argument to `compile` and so we track it via local flow. We could actually get fancy and track it via global flow using a second data-flow configuration, if we really wanted to.

Is this understanding accurate? If so, the two implementations are separate; concretely I think that `DirectRegex` does not find calls of the form `<compiled regex>.<ReMethod>` because `<compiled regex>` is not the `re` module.

Reply:
Trying to explain why I didn't see anything wrong here got to explain what `DirectRegex`'s `.getALocalSource()` was meant to do and realised what you spotted there 😆. Fixing it now :P thank you so much!